packages/commands/src/query/cctx.ts

```diff
@@ -0,0 +1,253 @@
+import axios from "axios";
+import { Command } from "commander";
+import { ethers } from "ethers";
+import EventEmitter from "eventemitter3";
+import { z } from "zod";
+
+import { CrossChainTx } from "../../../../types/cctx";
+
+/**
+ * Event map:
+ *  - `cctx` → emitted every polling round with the **entire** array so far.
+ */
+interface CctxEvents {
+  cctx: (allSoFar: CrossChainTx[]) => void;
+}
+
+export const cctxEmitter = new EventEmitter<CctxEvents>();
+
+const cctxOptionsSchema = z.object({
+  delay: z.coerce.number().int().positive().default(2000),
+  hash: z.string(),
+  rpc: z.string(),
+});
+
+type CctxOptions = z.infer<typeof cctxOptionsSchema>;
+
+interface CctxResponse {
+  CrossChainTxs: CrossChainTx[];
+}
+
+const sleep = (ms: number) => new Promise((resolve) => setTimeout(resolve, ms));
+
+const fetchCctx = async (
+  hash: string,
+  rpc: string
+): Promise<CrossChainTx[]> => {
+  const url = `${rpc}/zeta-chain/crosschain/inboundHashToCctxData/${hash}`;
+  const res = await axios.get<CctxResponse>(url, {
+    validateStatus: (s) => s === 200 || s === 404,
+  });
+  return res.status === 404 ? [] : res.data.CrossChainTxs;
+};
+
+/**
+ * True if the CCTX is still in-flight and may mutate on‑chain.
+ */
+const isPending = (tx: CrossChainTx): boolean =>
+  ["PendingOutbound", "PendingRevert"].includes(tx.cctx_status.status);
+
+/**
+ * Poll indefinitely until user terminates.
+ *
+ * • Hashes returning 404 are retried forever.
+ * • Every discovered `cctx.index` is queried **at least once** (to pull the
+ *   next hop) and queried again while its status remains pending.
+ */
+const gatherCctxs = async (
+  rootHash: string,
+  rpc: string,
+  delayMs = 2000
+): Promise<void> => {
+  // Latest copy of each CCTX keyed by index
+  const results = new Map<string, CrossChainTx>();
+
+  // Hashes we need to query in the upcoming round
+  const frontier = new Set<string>([rootHash]);
+
+  // Track which indexes we've *ever* queried so we still fetch each once
+  const queriedOnce = new Set<string>();
+
+  // eslint-disable-next-line no-constant-condition
+  while (true) {
+    const nextFrontier = new Set<string>();
+
+    await Promise.all(
+      [...frontier].map(async (hash) => {
+        try {
+          const cctxs = await fetchCctx(hash, rpc);
+
+          if (cctxs.length === 0) {
+            // Still 404 – keep trying
+            nextFrontier.add(hash);
+            return;
```

Comment:
Hashes returning 404 could be retried with exponential backoff and configurable "max retries"